internal/unpackinfo/filemode.go

```diff
@@ -0,0 +1,94 @@
+package unpackinfo
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+)
+
+type FileMode uint32
+
+const (
+	Empty FileMode = 0
+	// Dir represent a Directory.
+	Dir FileMode = 0040000
+	// Regular represent non-executable files.  Please note this is not
+	// the same as golang regular files, which include executable files.
+	Regular FileMode = 0100644
+	// Executable represents executable files.
+	Executable FileMode = 0100755
+	// Symlink represents symbolic links to files.
+	Symlink FileMode = 0120000
+)
+
+func NewFileMode(mode fs.FileMode) (FileMode, error) {
+	if mode.IsRegular() {
+		// disallow pipes, I/O, temporary files etc
+		if isCharDevice(mode) || isTemporary(mode) {
+			return Empty, fmt.Errorf("invalid file mode: %s", mode)
+		}
+
+		if isExecutable(mode) {
+			return Executable, nil
+		}
+
+		return Regular, nil
+	}
+
+	if mode.IsDir() {
+		return Dir, nil
+	}
+
+	if isSymlink(mode) {
+		return Symlink, nil
+	}
+
+	return Empty, fmt.Errorf("invalid file mode: %s", mode)
+}
+
+// Maps a FileMode integer to an fs.FileMode type, normalizing
+// permissions for regular and executable files.
+func (m FileMode) ToFsFileMode() (fs.FileMode, error) {
+	switch m {
+	case Regular:
+		return fs.FileMode(0644), nil
+	case Dir:
+		return fs.ModePerm | fs.ModeDir, nil
```

Comment:
This made me doubt myself and I ran in circles about it for a while, but! I think we actually want directories to be 755 (rwx r-x r-x), not 777 (world-writeable). Thoughts?

Reply:
In truth I followed the permissions set by [go-git](https://github.com/go-git/go-git/blob/9a6715aba7a88defd27aeba9ddb24cce9a9416c1/plumbing/filemode/filemode.go#L173). I wonder if there is a specific use case in git that requires dirs with r+w+x for all groups/users. I think for the purposes of Terraform configuration this would be irrelevant so we can get by with 0755. I will double check this to be sure.

Reply:
🤔 I wonder if there's a mask being applied somewhere later down the chain. Maybe do a real-world test and see what comes out; this is a case where I'm dubious about trusting what the code claims it's doing.